datafusion/expr-common/src/scalar.rs

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::{
+    array::{Array, ArrayRef},
+    datatypes::DataType,
+};
+use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
+
+#[derive(Clone, Debug)]
+pub struct Scalar {
+    value: ScalarValue,
+    data_type: DataType,
+}
+
+impl From<ScalarValue> for Scalar {
+    fn from(value: ScalarValue) -> Self {
+        Self {
+            data_type: value.data_type(),
+            value,
+        }
+    }
+}
+
+impl TryFrom<DataType> for Scalar {
+    type Error = DataFusionError;
+    fn try_from(value: DataType) -> Result<Self> {
+        Ok(Self {
+            value: ScalarValue::try_from(&value)?,
+            data_type: value,
+        })
+    }
+}
```

Comment:
Conversion from a data type to a value should be less implicit.
Can we call this function `new_null_of(type)`?

Reply:
I agree that the conversion should look less implicit so I'll move it to a dedicated function. I should mention that `ScalarValue` uses the same pattern:

https://github.com/apache/datafusion/blob/21ec3328428d4f27aa84abf7bd0c58332edcaed5/datafusion/common/src/scalar/mod.rs#L3354-L3361

and I adopted it to be consistent with that API.

Reply:
> I should mention that `ScalarValue` uses the same pattern:

yeah, i know. i still couldn't resist :)

Reply:
Sounds good! Updated 😄